### ISSUE TITLE
Missing programming ligatures

### ISSUE THREAD

**Issue by alekrutkowski** [intel/intel-one-mono#9]
Like those in e.g. Cascadia Code (https://github.com/microsoft/cascadia-code#font-features) and JuliaMono (https://juliamono.netlify.app/#contextual_and_stylistic_alternates_ligatures).

**Comment by MatthewTurrini** [comment]
Thanks for the feedback. Adding ligatures is something we may consider for a future release.

**Comment by CueXXIII** [comment]
Please don't. Or at least make them optional vi a 2nd font. Matthew Butterick has an informative blogpost about those ligatures which I fully agree with: https://practicaltypography.com/ligatures-in-programming-fonts-hell-no.html
tl;dr: They introduce ambiguities with their real unicode codepoints and they will be wrong in certain cases.

**Comment by Hussseinkizz** [comment]
Then without ligatures a few to no programmers will use this font, choose your audience, we want ligatures!!!

**Comment by Hussseinkizz** [comment]
> I would second the idea of maintaining dual version. Hence I incline to the idea of having ligatures.

Ok sure, that also works, and in some versions could something like what fira code iscript font does, making some things like imports, tag names and etc italic, I can share a shot of it if needed!

**Comment by alekrutkowski** [comment]
> Please don't. Or at least make them optional vi a 2nd font. Matthew Butterick has an informative blogpost about those ligatures which I fully agree with: https://practicaltypography.com/ligatures-in-programming-fonts-hell-no.html tl;dr: They introduce ambiguities with their real unicode codepoints and they will be wrong in certain cases.

Normally, you switch on or off ligatures in your text editor (e.g. VS Code) so you don't need two versions of a font.

**Comment by lone1y-51** [comment]
i want ligatures too.

**Comment by chrisdeeming** [comment]
This will be a hard sell for me without ligature support so support this 100%

**Comment by friscodelrosario** [comment]
I had already begun recommending IntelOne Mono as the greatest nerd font yet. I assumed ligatures were already in place, for thinking 'how could they not'?

Upvote, and I hope when you get around to implementation, the poorly-sighted folks reprise their parts.

**Comment by openendings** [comment]
> Normally, you switch on or off ligatures in your text editor (e.g. VS Code) so you don't need two versions of a font.

Seconding this: **ligatures can be disabled**. Fonts distributed in 'newer' formats like OpenType (`*.otf`) group ligatures etc. under 'features' (cf. ["stylistic sets"](https://learn.microsoft.com/en-us/typography/opentype/spec/features_pt#tag-ss01---ss20), see also [gsub](https://learn.microsoft.com/en-us/typography/opentype/spec/gsub) for starters), which can be independently toggled on/off. Recent IDEs and UIs allow toggling some of these, not others. E.g. Fira Code separates different ligatures [like so](https://github.com/tonsky/FiraCode/wiki/How-to-enable-stylistic-sets).

The maintainers would need to distribute some no-ligatures variations to ensure it wouldn't break older terminal clients, but for most other use cases it *ought* not to get in the way beyond a one-time settings change.

**Comment by baguse** [comment]
Upvote for this, I want to change my font to this but still waiting for the ligature support

**Comment by mahimairaja** [comment]
Wish to have ligature support

**Comment by sainnhe** [comment]
> Please don't. Or at least make them optional vi a 2nd font. Matthew Butterick has an informative blogpost about those ligatures which I fully agree with: https://practicaltypography.com/ligatures-in-programming-fonts-hell-no.html
> tl;dr: They introduce ambiguities with their real unicode codepoints and they will be wrong in certain cases.

You can always disable ligatures in your editor or terminal.

**Comment by YVVT** [comment]
Imho, a variant _with_ ligatures and a variant _without_ would satisfy both sides, just like JetBrains Mono does.

Trying to impose one's own opinion into another person seems inappropriate though -- That would lead to an infinite quarrle, as each side thinks they are right.

**Comment by YVVT** [comment]
> > Please don't. Or at least make them optional vi a 2nd font. Matthew Butterick has an informative blogpost about those ligatures which I fully agree with: https://practicaltypography.com/ligatures-in-programming-fonts-hell-no.html tl;dr: They introduce ambiguities with their real unicode codepoints and they will be wrong in certain cases.
> 
> Normally, you switch on or off ligatures in your text editor (e.g. VS Code) so you don't need two versions of a font.

However, unfortunately some IDEs (e.g. Microsoft Visual Studio) doesn't provide an option to switch ligatures on or off.

**Comment by thebodzio** [comment]
@YVVT From VS Code docs (https://code.visualstudio.com/docs/getstarted/settings):

```
// Configures font ligatures or font features. Can be either a boolean to enable/disable ligatures or a string for the value of the CSS 'font-feature-settings' property.
  "editor.fontLigatures": false,
```

Do not get me wrong: I think you are generally right when it comes to “some IDEs”, but still I do prefer ligatures to be a part of the font to be switched on or off. It is simply more convenient than having two separate fonts. That being said, how hard would it be to export a font file with degraded functionality (no ligatures embedded) in parallel to full font?

**Comment by abelguevaralanda** [comment]
Me too

**Comment by HTcbaldwin** [comment]
> However, unfortunately some IDEs (e.g. Microsoft Visual Studio) doesn't provide an option to switch ligatures on or off.

Sounds like you should log an issue on Microsoft's Visual Studio Github repo... :disappointed: Oh right you can't, it's closed source.

**Comment by YVVT** [comment]
> > However, unfortunately some IDEs (e.g. Microsoft Visual Studio) doesn't provide an option to switch ligatures on or off.
> 
> Sounds like you should log an issue on Microsoft's Visual Studio Github repo... :disappointed: Oh right you can't, it's closed source. 
> 

However, we can still post issues on Visual Studio Developer Community.

**Comment by zamaniafshar** [comment]
Please add ligatures sooner

**Comment by Tanhe** [comment]
Ligatures please. As others have mentioned in a separate font probably is the best solution for opt-in,

**Comment by FredShallcrass** [comment]
Hello programming ligature enthusiasts.

We are in the process of adding support for programming ligatures and box drawing characters to Intel One Mono. We are currently at a BETA stage which is a good time to get feedback on the character set and forms.

In this zip file is a beta build which you can install to try out the new character set. While you are trying out the ligatures, there are a few things to note:

Rendering – There are some rendering issues with overlapping glyphs in some of the programming ligatures. This will be addressed in the final build.

Feedback Timeline – In order to make sure your feedback is heard, please post any thoughts in this thread by the end of the week. (Sunday 17th of March)

Calt Feature – In this beta build, the ligatures will be activated by default through the calt feature. This will ensure they are easy to test, but in the final build the ligatures will be accessed though stylistic sets in order to ensure those who do not use ligatures can continue to have them off by default.

Thank you for your time and thoughts! 

[Intel One Mono Programming Ligature BETA.zip](https://github.com/intel/intel-one-mono/files/14590977/Intel.One.Mono.Programming.Ligature.BETA.zip)

**Comment by merlindru** [comment]
@FredShallcrass These ligatures are awesome! There's a huge set of them, you covered everything I could think of off the top of my head. Great work, really

Personally, i'd wish there's ligatures i could disable, like the `www` one (which confuses me when looking at it, esp because i don't see it often day-to-day)

It's really cool to see that things like `-----` becomes one long dash that still has "seams" where you can tell each character apart

or that `==>` works with infinitely many `=` chars

I also love that there are pairs that are just made to look better, like `:=` where the `:` changes in height, or `&&` and `::` where they move slightly closer together

Amazing work, thanks for creating & updating Intel One Mono <3

**Comment by FredShallcrass** [comment]
Thank you for the feedback! 

We can certainly separate out the www ligature from the main stylistic set. That glyph seems to be logically seperate too, so that works.

The fonts are being hinted right now and we will get started on a new build as soon as they are done, but we do not have a specific upload date yet.

**Comment by alphabitserial** [comment]
I'm looking forward to the next release with ligature support!

**Comment by FredShallcrass** [comment]
[Release 1.4.0](https://github.com/intel/intel-one-mono/releases/latest) includes box drawing glyphs and programming ligatures.

**Comment by rwmitchell** [comment]
I don't see any difference, I must be  doing something wrong.  I've tried using it with both iTerm2 and MacVim.  I thought '->' was a programming ligature, but it doesn't join into an arrow.  Other fonts do.

**Comment by RGB-es** [comment]
> I don't see any difference, I must be doing something wrong. I've tried using it with both iTerm2 and MacVim. I thought '->' was a programming ligature, but it doesn't join into an arrow. Other fonts do.

You need to activate the appropriate OpenType function. Programming ligatures are activated with the first stylistic set, (+ss01). On how to activate that option, I found this page from a different font project

https://github.com/tonsky/FiraCode/wiki/How-to-enable-stylistic-sets

Maybe Intel-one-mono need a similar page?
